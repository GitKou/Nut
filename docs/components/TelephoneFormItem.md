# TelephoneFormItem

固定电话表单项。

TelephoneInput 接收的 value 是格式如'0041-8323222'的字符串。
TelephoneFormItem 自带校验规则，区号位微 4 位数以内的数字字符串，号码为 8 位数以内的数字字符串。

## 代码示例

```tsx
import React from 'react';
import { Form, Button } from 'antd';
import { TelephoneFormItem } from '@lc-nut/components'; // 通过包名引入，而不是相对路径
export default () => {
  const onFinish = (values: any) => {
    console.log('Success:', values);
  };

  const onFinishFailed = (errorInfo: any) => {
    console.log('Failed:', errorInfo);
  };

  return (
    <Form
      initialValues={{ phone: 'gmm-123123' }}
      onFinish={onFinish}
      onFinishFailed={onFinishFailed}
    >
      <TelephoneFormItem name="phone" label="座机☎️" value />
      <Form.Item>
        <Button type="primary" htmlType="submit">
          Submit
        </Button>
      </Form.Item>
    </Form>
  );
};
```

<API src="../../packages/components/src/TelephoneFormItem/index.tsx" ></API>
