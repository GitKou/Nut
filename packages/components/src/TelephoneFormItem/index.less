.formItem {
  .selfError {
    background-color: pink;
  }
  .noSelfError {
    :global {
      :not(.ant-input-disabled):not(.ant-input-borderless).ant-input {
        border-color: rgba(22, 22, 64, 0.16);
      }
      :not(.ant-input-disabled):not(.ant-input-borderless).ant-input:focus {
        border-color: #5c85ff;
        box-shadow: 0 0 0 2px rgba(51, 95, 255, 0.2);
      }
    }
  }
}
