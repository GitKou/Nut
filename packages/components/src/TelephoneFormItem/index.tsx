import React, { useState } from 'react';
import type { FormItemProps } from 'antd';
import { Input, Form, Col, Row } from 'antd';
import cls from 'classnames';
import type { InputProps } from 'antd/es/input/Input';

import './index.less';

export type TelephoneInputProps = InputProps & {
  value?: string;
  defaultValue?: string;
  onChange?: (e?: string) => void;
  areaCodeHelp?: string;
  phoneNoHelp?: string;
};

export type TelephoneFormItemProps = FormItemProps<string> & {
  /** Input的props */
  inputProps?: TelephoneInputProps;
};

const getAreaCodeFromStr = (str?: string): string | undefined =>
  (str || '').split('-')[0];
const getPhoneNoFromStr = (str?: string): string | undefined =>
  (str || '').split('-')[1];

export enum TelephoneRules {
  请输入区号 = '请输入区号',
  区号不超过4位 = '区号不超过4位',
  区号应为数字 = '区号应为数字',
  请输入固定电话 = '请输入固定电话',
  固定电话应不超过8位 = '固定电话应不超过8位',
  固定电话应为数字 = '固定电话应为数字',
}
/**
 * 组合区号和固定电话号码，
 * 校验样式手动控制
 */
export function TelephoneInput({
  value,
  onChange,
  areaCodeHelp,
  phoneNoHelp,
  defaultValue,
  ...restInputProps
}: TelephoneInputProps) {
  const [areaCode, setAreaCode] = useState(
    getAreaCodeFromStr(value || defaultValue),
  );
  const [phoneNo, setPhoneNo] = useState(
    getPhoneNoFromStr(value || defaultValue),
  );

  const triggerChange = (changedValue: {
    areaCode?: string;
    phoneNo?: string;
  }) => {
    const newValue = {
      areaCode,
      phoneNo,
      ...changedValue,
    };
    const str = [newValue.areaCode, newValue.phoneNo].join('-');
    onChange?.(str === '-' ? undefined : str);
  };

  const handleAreaCodeChange = (e: { target: { value: string } }) => {
    setAreaCode(e.target.value);
    triggerChange({ areaCode: e.target.value });
  };
  const handlePhoneNoChange = (e: { target: { value: string } }) => {
    setPhoneNo(e.target.value);
    triggerChange({ phoneNo: e.target.value });
  };

  return (
    <Row>
      <Col span={8}>
        <Form.Item
          // name="areaCode"
          className={!areaCodeHelp ? 'no-self-error' : undefined}
          validateStatus={areaCodeHelp ? 'error' : undefined}
          help={areaCodeHelp}
        >
          <Input
            placeholder="请输入区号"
            value={areaCode}
            onChange={handleAreaCodeChange}
            {...restInputProps}
          />
        </Form.Item>
      </Col>
      <Col span={1}>
        <Row justify="center" align="middle" style={{ height: 36 }}>
          -
        </Row>
      </Col>
      <Col span={15}>
        <Form.Item
          // name="phoneNo"
          className={!phoneNoHelp ? 'no-self-error' : undefined}
          validateStatus={phoneNoHelp ? 'error' : undefined}
          help={phoneNoHelp}
        >
          <Input
            placeholder="请输入固定电话"
            value={phoneNo}
            onChange={handlePhoneNoChange}
            {...restInputProps}
          />
        </Form.Item>
      </Col>
    </Row>
  );
}

/** 区号是不超过4位的数字，固定是电话应不超过8位的数字 */
function TelephoneFormItem(props: TelephoneFormItemProps) {
  const { className, inputProps, ...restProps } = props;
  const [areaCodeHelp, setAreaCodeHelp] = useState<string>();
  const [phoneNoHelp, setPhoneNoHelp] = useState<string>();

  const formItemCls = cls('telephone-form-item', className);

  return (
    <Form.Item
      className={formItemCls}
      name="officePhone"
      label="固定电话"
      // 校验样式由内部input控制，外部不做控制
      validateStatus={undefined}
      help={''}
      rules={[
        {
          validator: (rules, value) => {
            if (!props.required && !value) {
              // 不是必填，且没有填值的时候不做校验
              setAreaCodeHelp(undefined);
              setPhoneNoHelp(undefined);
              return Promise.resolve();
            }

            const areaCode = getAreaCodeFromStr(value) || '';
            const phoneNo = getPhoneNoFromStr(value) || '';

            let areaCodeMsg = '';
            let phoneNoMsg = '';

            if (props.required && !areaCode) {
              areaCodeMsg = TelephoneRules.请输入区号;
            }
            if (props.required && !phoneNo) {
              areaCodeMsg = TelephoneRules.请输入固定电话;
            }

            if (areaCode.length > 4) {
              areaCodeMsg = TelephoneRules.区号不超过4位;
            }
            if (!/^[0-9]+$/.test(areaCode)) {
              areaCodeMsg = TelephoneRules.区号应为数字;
            }
            if (phoneNo.length > 8) {
              phoneNoMsg = TelephoneRules.固定电话应不超过8位;
            }
            if (!/^[0-9]+$/.test(phoneNo)) {
              phoneNoMsg = TelephoneRules.固定电话应为数字;
            }
            setAreaCodeHelp(areaCodeMsg);
            setPhoneNoHelp(phoneNoMsg);
            if (areaCodeMsg || phoneNoMsg) {
              return Promise.reject();
            }
            return Promise.resolve();
          },
        },
      ]}
      {...restProps}
    >
      <TelephoneInput
        areaCodeHelp={areaCodeHelp}
        phoneNoHelp={phoneNoHelp}
        {...inputProps}
      />
    </Form.Item>
  );
}

export default TelephoneFormItem;
