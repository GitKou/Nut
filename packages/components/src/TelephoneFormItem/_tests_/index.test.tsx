import { render, waitFor } from '@testing-library/react';
import { fireEvent, screen } from '@testing-library/dom';
import { Button, Form } from 'antd';
import renderer from 'react-test-renderer';
import '@testing-library/jest-dom';
import TelephoneFormItem, { TelephoneRules } from '../index';

const rulesValues = Object.values(TelephoneRules);
// rulesIdxs 期待不满足的规则下标
const msgsExpectToBeDisplayed = async (rulesIdxs) => {
  for (let i = 0; i < rulesValues.length; i++) {
    if (rulesIdxs.includes(rulesValues[i])) {
      // 当前输入满足条件idx
      await waitFor(() =>
        expect(
          screen.getByText(new RegExp(rulesValues[i])),
        ).toBeInTheDocument(),
      );
    } else {
      await waitFor(() =>
        expect(
          screen.getByText(new RegExp(rulesValues[i])),
        ).not.toBeInTheDocument(),
      );
    }
  }
};

describe('TelephoneFormItem', () => {
  it('renders correctly when there are multiple items in a Form', () => {
    const ele = renderer
      .create(
        <Form>
          <TelephoneFormItem label="tel1" name="tel1" />
          <TelephoneFormItem label="tel2" name="tel2" />
        </Form>,
      )
      .toJSON();
    expect(ele).toMatchSnapshot();
  });

  it('initialValue works in a Form', async () => {
    render(
      <Form initialValues={{ phone: 'gmm-123123' }}>
        <TelephoneFormItem label="固定电话" name="phone" />,
      </Form>,
    );
    const input1 = screen.getByPlaceholderText<HTMLInputElement>('请输入区号');
    const input2 =
      screen.getByPlaceholderText<HTMLInputElement>('请输入固定电话');

    expect(input1.value).toBe('gmm');
    expect(input2.value).toBe('123123');
  });

  it('shows no error when formItem is not required and the value is undefined or empty string', () => {
    render(
      <Form initialValues={{ phone2: '' }}>
        <TelephoneFormItem label="固定电话" name="phone" />
        <TelephoneFormItem label="固定电话" name="phone2" />
        <Button type="primary" htmlType="submit">
          Submit
        </Button>
      </Form>,
    );
    const submitter = screen.getByRole('button', {
      name: /submit/i,
    });

    fireEvent.click(submitter);

    msgsExpectToBeDisplayed([]);
  });

  it('shows required error messages when formItem is required and the value is undefined ', () => {
    render(
      <Form>
        <TelephoneFormItem label="固定电话" name="phone" required />
        <Button type="primary" htmlType="submit">
          Submit
        </Button>
      </Form>,
    );
    const submitter = screen.getByRole('button', {
      name: /submit/i,
    });

    fireEvent.click(submitter);

    msgsExpectToBeDisplayed([
      TelephoneRules.请输入区号,
      TelephoneRules.请输入固定电话,
    ]);
  });

  it('onchange callbacks correctly', () => {
    const onChange = jest.fn();
    render(
      <Form>
        <TelephoneFormItem
          label="固定电话"
          name="phone"
          required
          inputProps={{
            onChange,
          }}
        />
      </Form>,
    );
    const input1 = screen.getByPlaceholderText<HTMLInputElement>(/请输入区号/);
    const input2 =
      screen.getByPlaceholderText<HTMLInputElement>(/请输入固定电话/);
    fireEvent.change(input1, { target: { value: '12345' } });
    fireEvent.change(input2, { target: { value: 'gmm' } });
    expect(onChange).toHaveBeenCalledTimes(2);
  });

  it('warns when input.value does not pass specific validation rules.', async () => {
    render(
      <Form initialValues={{ phone: 'gmm-1231234444' }}>
        <TelephoneFormItem label="固定电话" name="phone" />
        <Button type="primary" htmlType="submit">
          Submit
        </Button>
      </Form>,
    );
    const input1 = screen.getByPlaceholderText<HTMLInputElement>(/请输入区号/);
    const input2 =
      screen.getByPlaceholderText<HTMLInputElement>(/请输入固定电话/);
    const submitter = screen.getByRole('button', {
      name: /submit/i,
    });

    fireEvent.click(submitter);
    // gmm-1231234444
    msgsExpectToBeDisplayed([
      TelephoneRules.区号应为数字,
      TelephoneRules.固定电话应不超过8位,
    ]);

    fireEvent.change(input1, { target: { value: '12345' } });
    fireEvent.change(input2, { target: { value: 'gmm' } });
    fireEvent.click(submitter);

    // 12345-gmm
    msgsExpectToBeDisplayed([
      TelephoneRules.区号不超过4位,
      TelephoneRules.固定电话应为数字,
    ]);

    fireEvent.change(input1, { target: { value: '0571' } });
    fireEvent.change(input2, { target: { value: '8322222' } });
    fireEvent.click(submitter);

    // 0571-8322222
    msgsExpectToBeDisplayed([]);
  });
});
