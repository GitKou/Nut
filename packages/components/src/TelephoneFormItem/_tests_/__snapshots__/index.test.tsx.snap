// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TelephoneFormItem renders correctly when there are multiple items in a Form 1`] = `
<form
  className="ant-form ant-form-horizontal"
  onReset={[Function]}
  onSubmit={[Function]}
>
  <div
    className="ant-row ant-form-item ant-form-item-with-help telephone-form-item"
    style={Object {}}
  >
    <div
      className="ant-col ant-form-item-label"
      style={Object {}}
    >
      <label
        className=""
        htmlFor="tel1"
        title="tel1"
      >
        tel1
      </label>
    </div>
    <div
      className="ant-col ant-form-item-control"
      style={Object {}}
    >
      <div
        className="ant-form-item-control-input"
      >
        <div
          className="ant-form-item-control-input-content"
        >
          <div
            className="ant-row"
            style={Object {}}
          >
            <div
              className="ant-col ant-col-8"
              style={Object {}}
            >
              <div
                className="ant-row ant-form-item no-self-error"
                style={Object {}}
              >
                <div
                  className="ant-col ant-form-item-control"
                  style={Object {}}
                >
                  <div
                    className="ant-form-item-control-input"
                  >
                    <div
                      className="ant-form-item-control-input-content"
                    >
                      <input
                        className="ant-input"
                        id="tel1"
                        onBlur={[Function]}
                        onChange={[Function]}
                        onFocus={[Function]}
                        onKeyDown={[Function]}
                        placeholder="请输入区号"
                        type="text"
                        value=""
                      />
                    </div>
                  </div>
                </div>
              </div>
            </div>
            <div
              className="ant-col ant-col-1"
              style={Object {}}
            >
              <div
                className="ant-row ant-row-center ant-row-middle"
                style={
                  Object {
                    "height": 36,
                  }
                }
              >
                -
              </div>
            </div>
            <div
              className="ant-col ant-col-15"
              style={Object {}}
            >
              <div
                className="ant-row ant-form-item no-self-error"
                style={Object {}}
              >
                <div
                  className="ant-col ant-form-item-control"
                  style={Object {}}
                >
                  <div
                    className="ant-form-item-control-input"
                  >
                    <div
                      className="ant-form-item-control-input-content"
                    >
                      <input
                        className="ant-input"
                        id="tel1"
                        onBlur={[Function]}
                        onChange={[Function]}
                        onFocus={[Function]}
                        onKeyDown={[Function]}
                        placeholder="请输入固定电话"
                        type="text"
                        value=""
                      />
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        className="ant-form-item-explain ant-form-item-explain-connected"
      >
        <div
          className="ant-show-help-item-appear ant-show-help-item-appear-start ant-show-help-item"
          role="alert"
          style={
            Object {
              "height": 0,
              "opacity": 0,
            }
          }
        >
          
        </div>
      </div>
    </div>
  </div>
  <div
    className="ant-row ant-form-item ant-form-item-with-help telephone-form-item"
    style={Object {}}
  >
    <div
      className="ant-col ant-form-item-label"
      style={Object {}}
    >
      <label
        className=""
        htmlFor="tel2"
        title="tel2"
      >
        tel2
      </label>
    </div>
    <div
      className="ant-col ant-form-item-control"
      style={Object {}}
    >
      <div
        className="ant-form-item-control-input"
      >
        <div
          className="ant-form-item-control-input-content"
        >
          <div
            className="ant-row"
            style={Object {}}
          >
            <div
              className="ant-col ant-col-8"
              style={Object {}}
            >
              <div
                className="ant-row ant-form-item no-self-error"
                style={Object {}}
              >
                <div
                  className="ant-col ant-form-item-control"
                  style={Object {}}
                >
                  <div
                    className="ant-form-item-control-input"
                  >
                    <div
                      className="ant-form-item-control-input-content"
                    >
                      <input
                        className="ant-input"
                        id="tel2"
                        onBlur={[Function]}
                        onChange={[Function]}
                        onFocus={[Function]}
                        onKeyDown={[Function]}
                        placeholder="请输入区号"
                        type="text"
                        value=""
                      />
                    </div>
                  </div>
                </div>
              </div>
            </div>
            <div
              className="ant-col ant-col-1"
              style={Object {}}
            >
              <div
                className="ant-row ant-row-center ant-row-middle"
                style={
                  Object {
                    "height": 36,
                  }
                }
              >
                -
              </div>
            </div>
            <div
              className="ant-col ant-col-15"
              style={Object {}}
            >
              <div
                className="ant-row ant-form-item no-self-error"
                style={Object {}}
              >
                <div
                  className="ant-col ant-form-item-control"
                  style={Object {}}
                >
                  <div
                    className="ant-form-item-control-input"
                  >
                    <div
                      className="ant-form-item-control-input-content"
                    >
                      <input
                        className="ant-input"
                        id="tel2"
                        onBlur={[Function]}
                        onChange={[Function]}
                        onFocus={[Function]}
                        onKeyDown={[Function]}
                        placeholder="请输入固定电话"
                        type="text"
                        value=""
                      />
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        className="ant-form-item-explain ant-form-item-explain-connected"
      >
        <div
          className="ant-show-help-item-appear ant-show-help-item-appear-start ant-show-help-item"
          role="alert"
          style={
            Object {
              "height": 0,
              "opacity": 0,
            }
          }
        >
          
        </div>
      </div>
    </div>
  </div>
</form>
`;

exports[`TelephoneFormItem renders correctly when there is a single TelephoneFormItem 1`] = `
<div
  className="ant-row ant-form-item ant-form-item-with-help telephone-form-item"
  style={Object {}}
>
  <div
    className="ant-col ant-form-item-label"
    style={Object {}}
  >
    <label
      className=""
      htmlFor="officePhone"
      title="固定电话"
    >
      固定电话
    </label>
  </div>
  <div
    className="ant-col ant-form-item-control"
    style={Object {}}
  >
    <div
      className="ant-form-item-control-input"
    >
      <div
        className="ant-form-item-control-input-content"
      >
        <div
          className="ant-row"
          style={Object {}}
        >
          <div
            className="ant-col ant-col-8"
            style={Object {}}
          >
            <div
              className="ant-row ant-form-item no-self-error"
              style={Object {}}
            >
              <div
                className="ant-col ant-form-item-control"
                style={Object {}}
              >
                <div
                  className="ant-form-item-control-input"
                >
                  <div
                    className="ant-form-item-control-input-content"
                  >
                    <input
                      className="ant-input"
                      id="officePhone"
                      onBlur={[Function]}
                      onChange={[Function]}
                      onFocus={[Function]}
                      onKeyDown={[Function]}
                      placeholder="请输入区号"
                      type="text"
                      value=""
                    />
                  </div>
                </div>
              </div>
            </div>
          </div>
          <div
            className="ant-col ant-col-1"
            style={Object {}}
          >
            <div
              className="ant-row ant-row-center ant-row-middle"
              style={
                Object {
                  "height": 36,
                }
              }
            >
              -
            </div>
          </div>
          <div
            className="ant-col ant-col-15"
            style={Object {}}
          >
            <div
              className="ant-row ant-form-item no-self-error"
              style={Object {}}
            >
              <div
                className="ant-col ant-form-item-control"
                style={Object {}}
              >
                <div
                  className="ant-form-item-control-input"
                >
                  <div
                    className="ant-form-item-control-input-content"
                  >
                    <input
                      className="ant-input"
                      id="officePhone"
                      onBlur={[Function]}
                      onChange={[Function]}
                      onFocus={[Function]}
                      onKeyDown={[Function]}
                      placeholder="请输入固定电话"
                      type="text"
                      value=""
                    />
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      className="ant-form-item-explain ant-form-item-explain-connected"
    >
      <div
        className="ant-show-help-item-appear ant-show-help-item-appear-start ant-show-help-item"
        role="alert"
        style={
          Object {
            "height": 0,
            "opacity": 0,
          }
        }
      >
        
      </div>
    </div>
  </div>
</div>
`;
